src/yaml.rs

```diff
@@ -656,6 +686,31 @@ impl Index<usize> for Yaml {
     }
 }
 
+impl IndexMut<usize> for Yaml {
+    fn index_mut(&mut self, idx: usize) -> &mut Yaml {
+        if let Some(v) = self.as_mut_vec() {
+            let yaml = v.get_mut(idx).unwrap();
+
+            // This transmute just casts the lifetime away. Since Rust only
+            // has SESE regions, this early return cannot be worked out and
+            // such that the borrow region of self includes the whole block.
+            // The explixit lifetimes in the function signature ensure that
+            // this is safe.
+            // From: https://github.com/image-rs/image-gif/blob/c814b40079a36af0b26adc36237f2c742d69c994/src/reader/decoder.rs#L227-L229
+            unsafe {
+                return mem::transmute::<&mut Yaml, &mut Yaml>(yaml);
+            }
+        }
+
+        if let Some(v) = self.as_mut_hash() {
+            let key = Yaml::Integer(i64::try_from(idx).unwrap());
+            v.get_mut(&key).unwrap()
+        } else {
+            panic!("Not a hash or verctor type");
+        }
+    }
```

Comment:
My other comment didn't get through.

You do not need the `unsafe` block if you only borrow `self` once.

While you're allowed to panic as per `IndexMut`'s interface, you're only allowed to do so if the index it out of range.
In this particular case, we have an additional condition in which we may panic which is the variant contained in `self` and this should be documented.

```suggestion
    /// Perform indexing if `self` is a sequence or a mapping.
    ///
    /// # Panics
    /// This function panics if the index given is out of range (as per [`IndexMut`]). If `self` i
    /// a [`Yaml::Array`], this is when the index is bigger or equal to the length of the
    /// underlying `Vec`. If `self` is a [`Yaml::Hash`], this is when the mapping sequence does no
    /// contain [`Yaml::Integer`]`(idx)` as a key.
    ///
    /// This function also panics if `self` is not a [`Yaml::Array`] nor a [`Yaml::Hash`].
    fn index_mut(&mut self, idx: usize) -> &mut Yaml {
        match self {
            Yaml::Array(sequence) => sequence.index_mut(idx),
            Yaml::Hash(mapping) => {
                let key = Yaml::Integer(i64::try_from(idx).unwrap());
                mapping.get_mut(&key).unwrap()
            }
            _ => panic!("Attempting to index but `self` is not a sequence nor a mapping"),
        }
    }
```